doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+
+| Num | Name                |
+| --- | ------------------- |
+| 0   | OK                  |
+| 1   | CANCELLED           |
+| 2   | UNKNOWN             |
+| 3   | INVALID_ARGUMENT    |
+| 4   | DEADLINE_EXCEEDED   |
+| 5   | NOT_FOUND           |
+| 6   | ALREADY_EXISTS      |
+| 7   | PERMISSION_DENIED   |
+| 8   | RESOURCE_EXHAUSTED  |
+| 9   | FAILED_PRECONDITION |
+| 10  | ABORTED             |
+| 11  | OUT_OF_RANGE        |
+| 12  | UNIMPLEMENTED       |
+| 13  | INTERNAL            |
+| 14  | UNAVAILABLE         |
+| 15  | DATA_LOSS           |
+| 16  | UNAUTHENTICATED     |
+
+Using the OK Status Code for a Call may only be decided by server applications.
+Library implementations must not "fabricate" an OK Status Code; it may only
+communicate an OK Status Code that was provided to it. While there may be
+additional restrictions on Status Code usage like those detailed in
+[statuscodes.md](statuscodes.md), those restrictions are more to provide a
+cohesive experience instead of a core, fundamental requirement.
+
+## Metadata
+
+Metadata has keys with associated values. Each key can have multiple values.
+Keys are unordered, but values for a key are ordered. Keys are case insensitive,
+and commonly canonicalized by making lower case. APIs are permitted to require a
+canonical representation and that representation may be different than the
+"lower case" representation mentioned here. A key can be for ASCII or binary
+values. If a key is for binary values, its name must be suffixed with "-bin".
+Otherwise it is for ASCII values.
+
+ASCII values are discouraged from having leading or trailing whitespace. If such
+a value contains leading or trailing whitespace, the whitespace may be stripped.
+Multiple ASCII values for the same key may be joined together with "," (a comma)
+as the delimiter and be considered semantically equivalent to the multi-value
+form.  However, such a transformation is lossy; an arbitrary ASCII value may not
+be split on comma and be assumed to be equivalent to a valid multi-value form
+for its key.
+
+The position of the Metadata in either Headers or Trailers is semantically
+important. Metadata from a Headers may not be moved to Trailers or vise-versa
+without additional knowledge of the individual key semantics.
+
+## Additional Features
+
+Although these are "additional" features, that does not make them unimportant.
+The are additional because they do not need to be represented as a fundamental
+gRPC protocol concept.
+
+### Deadline Propagation
+
+It is a common concept for RPCs to have a Deadline, a point in time by which the
+Call needs to complete before the client will give up. If the Client has a
+Deadline for the Call it should be included in the Request Headers. Calls are
+not required to have a Deadline. If a Call has a Deadline, both the client and
+server should track the Deadline and Cancel the Call when the Deadline is
+reached. When the client Cancels the Call, it should report a Trailers
```

Comment:
"it" is...the server?

If the client cancels, the server doesn't have a chance to send a trailer, at least in grpc-go.  The client cancels by sending RST_STREAM.  The stream is gone.

Or did you mean "when the client's deadline has expired"?

Reply:
The client library, which cancelled the call. The Trailers here is generated client-side. I'll replace the "it" to be more clear. Something needs to be here to make it clear that the "report" is "to the client application".